dns_server/src/crawler/core/mod.rs

```diff
@@ -0,0 +1,585 @@
+        NetworkingService,
+    },
+    peer_manager::global_ip::IsGlobalIp,
+    types::peer_address::PeerAddress,
+};
+
+/// How many outbound connection attempts can be made per heartbeat
+const MAX_CONNECTS_PER_HEARTBEAT: usize = 25;
+
+/// When the server drops the unreachable node address. Used for negative caching.
+const PURGE_UNREACHABLE_TIME: Duration = Duration::from_secs(3600);
+
+/// When the server drops the unreachable node address that was once reachable. This should take about a month.
+/// Such a long time is useful if the server itself has prolonged connectivity problems.
+pub const PURGE_REACHABLE_FAIL_COUNT: u32 = 35;
+
+/// Address state transition
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressTransition {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+/// Connection state of a potential node address (outbound only)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressState {
+    Disconnected {
+        /// The number of consecutive failed connection attempts.
+        /// New connection attempts are made after a progressive backoff time.
+        fail_count: u32,
+    },
+    Connecting {
+        /// Same as in the Disconnected state
+        fail_count: u32,
+    },
+    Connected {},
+}
+
+/// Additional state of a potential node address
+struct AddressData {
+    /// Connection state
+    state: AddressState,
+
+    /// Last time when the connection state was updated
+    state_updated_at: Duration,
+
+    /// Whether the address was reachable at least once.
+    /// Addresses that were once reachable are stored in the DB.
+    was_reachable: bool,
+
+    /// Whether the address was added from the command line
+    user_added: bool,
+}
+
+#[derive(Clone)]
+pub struct CrawlerCoreConfig {
+    /// Default p2p port on which nodes normally accept inbound connections
+    pub default_p2p_port: u16,
+}
+
+pub struct CrawlerCore<N: NetworkingService, C> {
+    /// Time of some monotonic timer, started from 0
+    now: Duration,
+
+    /// Crawler config
+    config: CrawlerCoreConfig,
+
+    /// Chain config
+    chain_config: Arc<ChainConfig>,
+
+    /// Map of all known addresses (including currently unreachable)
+    addresses: BTreeMap<N::Address, AddressData>,
+
+    /// Map of all currently connected outbound peers
+    outbound_peers: BTreeMap<N::PeerId, N::Address>,
+
+    /// CoreCallback implementation
+    _phantom: PhantomData<C>,
+}
+
+pub enum CoreEvent<N: NetworkingService> {
+    Timer { period: Duration },
+    Connectivity { event: ConnectivityEvent<N> },
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum CoreCommand<N: NetworkingService> {
+    DnsAddAddress(IpAddr),
+    DnsDelAddress(IpAddr),
+
+    DbStoreAddress(N::Address),
+    DbEraseAddress(N::Address),
+
+    Connect(N::Address),
+    Disconnect(N::PeerId),
+    SendResponse(N::PeerId, PeerManagerMessage),
+}
+
+pub trait CoreCallback<N: NetworkingService> {
+    fn exec(&mut self, command: CoreCommand<N>);
```

Comment:
Why not use something like `FnMut()` and alias it, instead of creating a trait?

Reply:
To be clear, I don't mind. Was just wondering.

Reply:
I started with `FnMut`, but then decided that a separate trait would be easier to use.